File: python/run-length-encoding/run_length_encoding.py
def decode(string):
  res = []
  i = 0
  while i < len(string):
    char = string[i]
    if char.isdigit():
      num = char
      while string[i+1].isdigit():
        num += string[i+1]
        i += 1
      res.append(int(num) * string[i+1])
      i += 2
    else:
      res.append(char)
      i += 1
  
  return ''.join(res)


def encode(string):
  res = []
  count = 0
  for i, char in enumerate(string):
    count += 1
    if i == len(string) - 1 or char != string[i+1]:
      if count != 1:
        res.append(str(count))
      res.append(char)
      count = 0

  return ''.join(res)

File: kotlin/run-length-encoding/src/main/kotlin/RunLengthEncoding.kt
object RunLengthEncoding {

    fun encode(input: String): String {
        var count = 1
        return input.foldIndexed("") { i, acc, c  ->
            var letters = ""
            if (c == input.getOrNull(i + 1)) {
                count++
            } else {
                letters += if (count == 1) c else "$count$c"
                count = 1
            }
            acc + letters
        }
    }

    fun decode(input: String): String {
        var count = ""
        return input.fold("") { acc, c -> 
            var letters = ""
            if (c.isDigit()) {
                count += c
            } else {
                letters += if (count != "") c.toString().repeat(count.toInt()) else c
                count = ""
            }
            acc + letters
        }
    }
}
